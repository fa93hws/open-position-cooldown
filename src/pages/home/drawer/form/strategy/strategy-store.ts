import type { ComponentType } from 'react';
import { makeObservable, observable, action, computed } from 'mobx';

import { ExposedInputProps } from '@ui/text-input/input';
import { createLine } from './line/line';
import { LineStore } from './line/line-store';

export class StrategyStore {
  strategies: {
    Component: ComponentType<ExposedInputProps>;
    store: LineStore;
  }[] = [];

  shouldShowRemove = false;

  constructor() {
    makeObservable(this, {
      strategies: observable.deep,
      shouldShowRemove: observable.ref,
      addStrategy: action,
      removeStrategy: action,
      StrategyLines: computed,
      strategyPlan: computed,
      strategyErrors: computed,
      setRemoveVisibility: action,
    });
  }

  get StrategyLines(): ComponentType<ExposedInputProps>[] {
    return this.strategies.map((strategy) => strategy.Component);
  }

  get strategyPlan() {
    return this.strategies.map(({ store }) => ({
      quantity: store.quantity,
      price: store.price,
    }));
  }

  get strategyErrors() {
    return this.strategies.map(({ store }) => store.hasError);
  }

  addStrategy() {
    const [Component, store] = createLine();
    this.strategies.push({ Component, store });
    return store;
  }

  removeStrategy(idx: number) {
    this.strategies.splice(idx, 1);
  }

  setRemoveVisibility(val: boolean) {
    this.shouldShowRemove = val;
  }

  startValidate() {
    this.strategies.forEach(({ store }) => store.startValidate());
  }
}
