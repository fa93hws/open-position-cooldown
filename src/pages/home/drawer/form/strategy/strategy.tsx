import * as React from 'react';
import { observer } from 'mobx-react';
import { Box, Typography, List, ListItem, IconButton } from '@material-ui/core';
import { makeStyles, withTheme, WithTheme } from '@material-ui/core/styles';
import IndeterminateCheckBoxIcon from '@material-ui/icons/IndeterminateCheckBox';

import { createInput, ExposedInputProps } from '@ui/text-input/input';
import { nonEmpty, mustBeNumber } from '@ui/text-input/validator';
import { ListControl } from '../list-control/list-control';
import { useInputStyles } from './line/line';
import { StrategyStore } from './strategy-store';

type LineItemProps = WithTheme & {
  Line: React.ComponentType;
  shouldShowRemove: boolean;
  idx: number;
  onRemoveClick(idx: number): void;
};

const useStyles = makeStyles((theme) => ({
  iconButton: {
    padding: 0,
    marginLeft: theme.spacing(1),
  },
}));

const LineItem = React.memo(
  withTheme((props: LineItemProps) => {
    const styles = useStyles(props.theme);
    const onRemoveClick = React.useCallback(
      () => props.onRemoveClick(props.idx),
      [props.idx, props.onRemoveClick],
    );
    return (
      <ListItem alignItems="center" dense disableGutters>
        <props.Line />
        {props.shouldShowRemove && (
          <IconButton
            color="secondary"
            className={styles.iconButton}
            onClick={onRemoveClick}
            title="remove"
          >
            <IndeterminateCheckBoxIcon />
          </IconButton>
        )}
      </ListItem>
    );
  }),
);

type StrategySectionProps = WithTheme & {
  ListControlImpl: React.ComponentType;
  CurrentQuantityInput: React.ComponentType<ExposedInputProps>;
  ShitPriceInput: React.ComponentType<ExposedInputProps>;
  Lines: readonly React.ComponentType[];
  shouldShowRemove: boolean;
  onRemoveClick(idx: number): void;
};

export const Strategy = React.memo(
  withTheme((props: StrategySectionProps) => {
    const inputStyles = useInputStyles(props.theme);
    return (
      <Box>
        <Box display="flex">
          <Typography variant="h6" component="h4">
            之后暴跌怎么办
          </Typography>
          <Box ml="auto">
            <props.ListControlImpl />
          </Box>
        </Box>
        <Box display="flex" alignItems="center" my={1}>
          <Typography variant="body1" component="span">
            当前价格买入
          </Typography>
          <Box width={80} mx={1}>
            <props.CurrentQuantityInput inputClasses={inputStyles} />
          </Box>
          <Typography variant="body1" component="span">
            股
          </Typography>
        </Box>
        <List dense disablePadding>
          {props.Lines.map((Line, idx) => (
            <LineItem
              key={idx}
              idx={idx}
              Line={Line}
              shouldShowRemove={props.shouldShowRemove}
              onRemoveClick={props.onRemoveClick}
            />
          ))}
        </List>
        <Box display="flex" alignItems="center" my={1}>
          <Typography variant="body1" component="span">
            跌至
          </Typography>
          <Box width={100} mx={1}>
            <props.ShitPriceInput inputClasses={inputStyles} />
          </Box>
          <Typography variant="body1" component="span">
            以下时:&nbsp;&nbsp;
          </Typography>
          <Typography variant="body1" component="span" color="secondary">
            没辙了
          </Typography>
        </Box>
      </Box>
    );
  }),
);

export function createStrategyPanel(): [React.ComponentType, StrategyStore] {
  const [CurrentQuantityInput, currentPriceStore] = createInput([
    nonEmpty,
    mustBeNumber,
  ]);
  const [ShitPriceInput, shitPriceStore] = createInput([
    nonEmpty,
    mustBeNumber,
  ]);
  const store = new StrategyStore({ currentPriceStore, shitPriceStore });
  store.reset();
  const onAddClick = () => store.addStrategy();
  const onSwitchChange = (val: boolean) => store.setRemoveVisibility(val);

  const ListControlImpl = observer(() => (
    <ListControl
      removeChecked={store.shouldShowRemove}
      onRemoveChange={onSwitchChange}
      onAddClick={onAddClick}
    />
  ));
  const onRemoveClick = (idx: number) => store.removeStrategy(idx);

  const Component = observer(() => (
    <Strategy
      ListControlImpl={ListControlImpl}
      Lines={store.StrategyLines}
      CurrentQuantityInput={CurrentQuantityInput}
      ShitPriceInput={ShitPriceInput}
      shouldShowRemove={store.shouldShowRemove}
      onRemoveClick={onRemoveClick}
    />
  ));
  return [Component, store];
}
