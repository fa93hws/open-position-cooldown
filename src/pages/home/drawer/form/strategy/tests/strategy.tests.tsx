import * as React from 'react';
import { shallow } from 'enzyme';

import { Strategy, createStrategyPanel } from '../strategy';

describe('Strategy', () => {
  it('renders the folowing', () => {
    const ListControlImpl = () => <div>ListControlImpl</div>;
    expect(
      <Strategy ListControlImpl={ListControlImpl} />,
    ).toMatchRenderedSnapshot();
  });

  it('is abled to be mounted', () => {
    const StrategyPanel = createStrategyPanel();
    expect(() => shallow(<StrategyPanel />)).not.toThrow();
  });
});
