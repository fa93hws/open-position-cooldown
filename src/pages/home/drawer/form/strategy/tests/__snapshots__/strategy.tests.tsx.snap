// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Strategy renders the strategy panel with two lines 1`] = `
<div
  class="MuiBox-root MuiBox-root-10"
>
  <div
    class="MuiBox-root MuiBox-root-11"
  >
    <h4
      class="MuiTypography-root MuiTypography-h6"
    >
      建仓策略
    </h4>
    <div
      class="MuiBox-root MuiBox-root-12"
    >
      <div>
        ListControlImpl
      </div>
    </div>
  </div>
  <div
    class="MuiBox-root MuiBox-root-13"
  >
    <span
      class="MuiTypography-root MuiTypography-body1"
    >
      当前价格买入
    </span>
    <div
      class="MuiBox-root MuiBox-root-14"
    >
      <div>
        ShitPriceInput
      </div>
    </div>
    <span
      class="MuiTypography-root MuiTypography-body1"
    >
      股
    </span>
  </div>
  <ul
    class="MuiList-root MuiList-dense"
  >
    <li
      class="MuiListItem-root MuiListItem-dense"
    >
      <div>
        strategy 1
      </div>
    </li>
    <li
      class="MuiListItem-root MuiListItem-dense"
    >
      <div>
        strategy 2
      </div>
    </li>
  </ul>
  <div
    class="MuiBox-root MuiBox-root-16"
  >
    <span
      class="MuiTypography-root MuiTypography-body1"
    >
      跌至
    </span>
    <div
      class="MuiBox-root MuiBox-root-17"
    >
      <div>
        ShitPriceInput
      </div>
    </div>
    <span
      class="MuiTypography-root MuiTypography-body1 MuiTypography-colorSecondary"
    >
      妈的! 加不起！
    </span>
  </div>
</div>
`;

exports[`Strategy renders the strategy panel with two lines and remove icon button 1`] = `
<div
  class="MuiBox-root MuiBox-root-18"
>
  <div
    class="MuiBox-root MuiBox-root-19"
  >
    <h4
      class="MuiTypography-root MuiTypography-h6"
    >
      建仓策略
    </h4>
    <div
      class="MuiBox-root MuiBox-root-20"
    >
      <div>
        ListControlImpl
      </div>
    </div>
  </div>
  <div
    class="MuiBox-root MuiBox-root-21"
  >
    <span
      class="MuiTypography-root MuiTypography-body1"
    >
      当前价格买入
    </span>
    <div
      class="MuiBox-root MuiBox-root-22"
    >
      <div>
        ShitPriceInput
      </div>
    </div>
    <span
      class="MuiTypography-root MuiTypography-body1"
    >
      股
    </span>
  </div>
  <ul
    class="MuiList-root MuiList-dense"
  >
    <li
      class="MuiListItem-root MuiListItem-dense"
    >
      <div>
        strategy 1
      </div>
      <button
        class="MuiButtonBase-root MuiIconButton-root makeStyles-iconButton-15 MuiIconButton-colorSecondary"
        id="remove-0"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-2 10H7v-2h10v2z"
            />
          </svg>
        </span>
      </button>
    </li>
    <li
      class="MuiListItem-root MuiListItem-dense"
    >
      <div>
        strategy 2
      </div>
      <button
        class="MuiButtonBase-root MuiIconButton-root makeStyles-iconButton-15 MuiIconButton-colorSecondary"
        id="remove-1"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-2 10H7v-2h10v2z"
            />
          </svg>
        </span>
      </button>
    </li>
  </ul>
  <div
    class="MuiBox-root MuiBox-root-23"
  >
    <span
      class="MuiTypography-root MuiTypography-body1"
    >
      跌至
    </span>
    <div
      class="MuiBox-root MuiBox-root-24"
    >
      <div>
        ShitPriceInput
      </div>
    </div>
    <span
      class="MuiTypography-root MuiTypography-body1 MuiTypography-colorSecondary"
    >
      妈的! 加不起！
    </span>
  </div>
</div>
`;

exports[`Strategy renders the strategy panel without lines 1`] = `
<div
  class="MuiBox-root MuiBox-root-3"
>
  <div
    class="MuiBox-root MuiBox-root-4"
  >
    <h4
      class="MuiTypography-root MuiTypography-h6"
    >
      建仓策略
    </h4>
    <div
      class="MuiBox-root MuiBox-root-5"
    >
      <div>
        ListControlImpl
      </div>
    </div>
  </div>
  <div
    class="MuiBox-root MuiBox-root-6"
  >
    <span
      class="MuiTypography-root MuiTypography-body1"
    >
      当前价格买入
    </span>
    <div
      class="MuiBox-root MuiBox-root-7"
    >
      <div>
        ShitPriceInput
      </div>
    </div>
    <span
      class="MuiTypography-root MuiTypography-body1"
    >
      股
    </span>
  </div>
  <ul
    class="MuiList-root MuiList-dense"
  />
  <div
    class="MuiBox-root MuiBox-root-8"
  >
    <span
      class="MuiTypography-root MuiTypography-body1"
    >
      跌至
    </span>
    <div
      class="MuiBox-root MuiBox-root-9"
    >
      <div>
        ShitPriceInput
      </div>
    </div>
    <span
      class="MuiTypography-root MuiTypography-body1 MuiTypography-colorSecondary"
    >
      妈的! 加不起！
    </span>
  </div>
</div>
`;
